wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/URLFileContentProvider.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.factory.server.urlfactory;
+
+import static java.lang.String.format;
+
+import java.io.IOException;
+import java.net.URI;
+import org.eclipse.che.api.devfile.server.FileContentProvider;
+
+/**
+ * A simple implementation of the FileContentProvider that merely uses the function resolve relative
+ * paths and {@link URLFetcher} for retrieving the content, handling common error cases.
+ */
+public class URLFileContentProvider implements FileContentProvider {
+
+  private final URI devfileLocation;
+  private final URLFetcher urlFetcher;
+
+  public URLFileContentProvider(URI devfileLocation, URLFetcher urlFetcher) {
+    this.devfileLocation = devfileLocation;
+    this.urlFetcher = urlFetcher;
+  }
+
+  @Override
+  public String fetchContent(String fileName) throws IOException {
+    URI fileUrl = devfileLocation.resolve(fileName);
+    try {
+      return urlFetcher.fetch(fileUrl.toString());
+    } catch (IOException e) {
+      throw new IOException(
+          format(
+              "Failed to fetch a file %s as relative to devfile %s from URL %s. Make sure the"
+                  + " location of the devfile points to the raw content of it (e.g. on Github or other"
+                  + " git repository make sure to use a link to the raw content of the devfile, not"
+                  + " a link to the webpage showing it). If obtaining such URL is not possible, make"
+                  + " sure you have installed a factory resolver plugin for given repository type"
+                  + " to your Che server instance.",
+              fileName, devfileLocation, fileUrl),
```

Comment:
I think it could be useful to see an original error message as well, like it can show 403 which indicates token absence and can help a user to understand a cause. WDYT?

Reply:
I think it is better to include an original error message and cause.
I do not know component which would propagate causes' errors to the user, but the cause can be useful as well if an exception is logged.

Reply:
Makes sense.